spec/money/allocation_spec.rb

```diff
@@ -0,0 +1,118 @@
+        expect(described_class.new(BigDecimal(100), 5, false).generate).to eq([20, 20, 20, 20, 20])
+      end
+
+      it 'splits the amount into equal fractions' do
+        expect(described_class.new(BigDecimal(5), 2, false).generate).to eq([2.5, 2.5])
+        expect(described_class.new(BigDecimal(5), 4, false).generate).to eq([1.25, 1.25, 1.25, 1.25])
+      end
+
+      it 'handles splits into repeating decimals' do
+        amount = BigDecimal(100)
+        parts = described_class.new(amount, 3, false).generate
+
+        expect(parts).to eq([
+          BigDecimal('33.3333333333333333335'),
+          BigDecimal('33.3333333333333333335'),
+          BigDecimal('33.333333333333333333')
+        ])
+        expect(parts.inject(0, :+)).to eq(amount)
+      end
+    end
+  end
+
+  describe 'given array as argument' do
+    it 'raises an error when invalid argument is given' do
+      expect { described_class.new(100, []) }.to raise_error(ArgumentError)
+    end
+
+    context 'whole amounts' do
+      it 'returns the amount when array contains only one element' do
+        expect(described_class.new(100, [1]).generate).to eq([100])
+        expect(described_class.new(100, [5]).generate).to eq([100])
+      end
+
+      it 'splits the amount into whole parts respecting the order' do
+        expect(described_class.new(100, [1, 1]).generate).to eq([50, 50])
+        expect(described_class.new(100, [1, 1, 2]).generate).to eq([25, 25, 50])
+        expect(described_class.new(100, [7, 3]).generate).to eq([70, 30])
+      end
+
+      it 'accepts floats as arguments' do
+        expect(described_class.new(100, [1.0, 1.0]).generate).to eq([50, 50])
+        expect(described_class.new(100, [0.1, 0.1, 0.2]).generate).to eq([25, 25, 50])
+        expect(described_class.new(100, [0.07, 0.03]).generate).to eq([70, 30])
+      end
+
+      it 'does not loose pennies' do
+        expect(described_class.new(10, [1, 1, 2]).generate).to eq([3, 2, 5])
+        expect(described_class.new(100, [1, 1, 1]).generate).to eq([34, 33, 33])
+      end
+    end
+
+    context 'fractional amounts' do
+      it 'returns the amount when array contains only one element' do
+        expect(described_class.new(BigDecimal(100), [1], false).generate).to eq([100])
+        expect(described_class.new(BigDecimal(100), [5], false).generate).to eq([100])
+      end
+
+      it 'splits the amount into whole parts respecting the order' do
+        expect(described_class.new(BigDecimal(100), [1, 1], false).generate).to eq([50, 50])
+        expect(described_class.new(BigDecimal(100), [1, 1, 2], false).generate).to eq([25, 25, 50])
+        expect(described_class.new(BigDecimal(100), [7, 3], false).generate).to eq([70, 30])
+      end
+
+      it 'splits the amount proportionally to the given parts' do
+        expect(described_class.new(BigDecimal(10), [1, 1, 2], false).generate).to eq([2.5, 2.5, 5])
```

Comment:
Since it's hooked into `Money`, the `whole_amounts` is coupled to the type of the amount. I'll add a spec that you've suggested to make it intentional, but if you think like there's a point in treating the amount like a float when `whole_amounts` is `false`, I can make it work as well.

Reply:
I think what you made is perfectly fine! I just stumbled upon the corner case — which is kinda even an "outside (out of scope) corner" 😉